Improve NVIC performance

kartben opened the pull request:
This PR implements several performance improvements in the NVIC emulation. There is a performance improvement of over 10x for the Azure RTOS ThreadX applications I have tested this against, and I am confident other RTOSes will see similar improvements.
See commit messages for detailed changelog!

/cc @PiotrZierhoffer.

PiotrZierhoffer commented:
Thanks for the contribution! We are running our internal tests on them and we will get back to you with feedback,

PiotrZierhoffer commented on src/Emulator/Cores/Arm-M/CortexM.cs:
```diff
@@ -204,7 +205,7 @@ private void OnPRIMASKWrite(int value)
         private int PendingMaskedIRQ()
         {
-            return nvic.MaskedInterruptPresent.WaitOne(0) ? 1 : 0;
+            return nvic.MaskedInterruptPresent ? 1 : 0;
```
This is a little bit controversial - I need to trace down why was it here in the first place. It looks ok now, but just want to be 100% sure

kartben replied:
Agreed this definitely looks like a substantial/risky change. I really couldn't get my head around why this might have been here in the first place so unfortunately, I won't be of much help :) What I do know is that the overhead of the ManualResetEvent is not negligible.

PiotrZierhoffer replied:
Ok, we believe it's good to be moved to `bool`. `MRE` was used back in the days when timers had their own threads, but this is no longer the case!

PiotrZierhoffer reviewed (changes requested):
Can you please prepend the commit names with `[GH#30] nvic:`?

kartben commented on src/Emulator/Cores/Arm-M/NVIC.cs:
```diff
@@ -29,6 +30,7 @@ public NVIC(Machine machine, long systickFrequency = 50 * 0x800000, byte priorit
         {
             priorities = new byte[IRQCount];
             activeIRQs = new Stack<int>();
+            pendingIRQs = new HashSet<int>();
```
+1

kartben replied:
addressed in latest version of the branch

kartben commented:
> Can you please prepend the commit names with `[GH#30] nvic:`?

Will do. If that's OK I suggest waiting until the very end of the review process as if I rewrite history in my branch I will likely screw up the inlined comments.

kartben commented on src/Emulator/Cores/Arm-M/NVIC.cs:
```diff
@@ -539,15 +548,18 @@ private int FindPendingInterrupt()
                         this.NoisyLog("IRQ {0} preempts {1}.", result, activeIRQs.Peek());
                     }
                 }
-                IRQ.Set(!PRIMASK && result != SpuriousInterrupt);
```
I came at this thinking that only AcknowledgeIRQ() would be reponsible for eventually unsetting the IRQ, and that FindPendingInterrupt may only ever be in situations where it'd set  IRQ from false->true or true->true, if that makes sense?

PiotrZierhoffer replied:
I am not yet sure why does the test software not go the usual way. I will look more into this, as it is quite surprising.

kartben replied:
just curious if you have been able to dig into this one more? BTW I think I lost track of what are the different changes you think should be done, I've taken note of the change from HashSet to SortedSet, but not 100% sure what the other ones should be :)

PiotrZierhoffer replied:
Yes, I did spend some time on this. Unfortunately it's not super trivial to pinpoint, as the fail in my test case does not happen immediately.

My best bet at the moment is that this fail is related to BASEPRI being set. I am not 100% sure yet, but it does look sensible.

Honestly speaking I'm inclined towards preserving the previous logic instead of digging deeper here. I don't think it should influence your performance very much.

kartben replied:
Would it be possible to share the tests that are failing, by any chance? The impact on performance is about 2x so it is still something that I hope can be fixed eventually (that is, unless the logic is not right in the first place, and IRQ should indeed be forcefully unset as part of FindPendingInterrupt)

PiotrZierhoffer replied:
Oh, that's very significant! I wasn't aware this change had such an impact.

Well, there are three possible cases here:
1) the test is wrong
2) the algorithm in NVIC is wrong in other places
3) your speed-up is due to interrupts appearing not often enough.

I think we can rule out option 1 with, let's say, 80% confidence. We use a Zephyr-based test, it's not really fancy.
The test is here: https://github.com/renode/renode/blob/master/tests/platforms/NRF52840.robot#L88
I'll try to dig up the sources for this test, but's it's a generic Zephyr sample with some configuration.

Option 2 is possible of course.

Option 3: I'm not sure. The application might not even notice that some interrupts are appearing less often.

PiotrZierhoffer replied:
You can see it run by `./test.sh --show-log -f "*NRF52840.Should Run ZephyrRTOS Shell On UART*" tests/platforms/NRF52840.robot` (unfortunately it requires ctrl+c after running)

PiotrZierhoffer commented:
Of course, I would actually prefer if we kept the history incremental until the very end

PiotrZierhoffer commented on src/Emulator/Cores/Arm-M/NVIC.cs:
```diff
@@ -539,15 +548,18 @@ private int FindPendingInterrupt()
                 else
                 {
-                    maskedInterruptPresent.Reset();
+                    maskedInterruptPresent = false;
```
if we revert the `IRQ.Set` part, then this will become just `maskedInterruptPresent = result != SpuriousInterrupt;`

kartben commented on src/Emulator/Cores/Arm-M/NVIC.cs:
```diff
@@ -268,6 +270,7 @@ public void Reset()
                 priorities[i] = 0x00;
             }
             activeIRQs.Clear();
+            pendingIRQs.Clear();
```
addressed in latest version of the branch

kartben commented:
> Can you please prepend the commit names with `[GH#30] nvic:`?

done!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renode/renode-infrastructure?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/renode/renode-infrastructure?pullRequest=30) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: kartben<br/>:x: PiotrZierhoffer<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renode/renode-infrastructure?pullRequest=30) it.</sub>

PiotrZierhoffer reviewed (approved):
I applied the IsCandidate changes myself

PiotrZierhoffer commented:
I have added some minor formatting changes, moved one thing between the commits and added one commit on top. It should work for you just fine, but please double check.

I'm running internal tests right now, will merge when we're green.

kartben commented:
> I have added some minor formatting changes, moved one thing between the commits and added one commit on top. It should work for you just fine, but please double check.

Thanks! I just tested it and it seems to work just fine :)
